### ISSUE TITLE
Missing metrics are not filled from other serves in server group #2

### ISSUE THREAD

**Issue by matejzero** [jacksontj/promxy#337]
Hello,

I first wanted to add my case under https://github.com/jacksontj/promxy/issues/325, but it seems our cases are not exactly the same as the end result is different, but the title matches my case. Feel free to change it if you can think of a better one.

My case:
I have one Prometheus in DC1 and one Prometheus in DC2. Same goes for VictoriaMetrics. Prometheus is doing a cross-dc scraping and feeding data to local VictoriaMetrics server. I'm using Promxy to deduplicate and fill data.

Schema:
![Screenshot 2020-08-18 at 11 06 57](https://user-images.githubusercontent.com/4689252/90493698-f738b000-e142-11ea-8162-4862a760a5de.png)
 
Today I upgraded my VM instances and while one was restarting, I got a hole in my graph even though the other one was able to provide data. But the end result was a graph with holes in it.

Example:
![Screenshot 2020-08-18 at 11 17 29](https://user-images.githubusercontent.com/4689252/90495040-6c58b500-e144-11ea-86dc-6c64e6f33200.png)

The first graph is the restarted victoriametric server, the second one was running at the time and the third one is the result from Promxy.

Promxy config (my scraping interval is 10s):
```yaml
global:
  evaluation_interval: 5s
  external_labels:
    source: promxy
rule_files:
  - "*rule"
promxy:
  server_groups:
    - static_configs:
        - targets:
          - vm-dc1:8428
          - vm-dc2:8428
      labels:
        sg: prometheus
      anti_affinity: 10s
      remote_read: true
      remote_read_path: api/v1/read
      query_params:
        nocache: 1
      http_client:
        dial_timeout: 1s
        tls_config:
          insecure_skip_verify: true
```

Is this a bug or iz my config / understanding wrong?

**Comment by jacksontj** [comment]
This _looks_ like a bug! Would you be able to provide either a tcpdump or the trace log output of the query with the hole in it? Given the setup you described I'd expect promxy to stitch the data together since they both are configured in the same group and have "the same" data.

**Comment by matejzero** [comment]
Sure!

Paste of the query with the holes (last graph) is on https://pastebin.com/tTxRdnbu

**Comment by jacksontj** [comment]
@matejzero sorry about the long delay! So I did some digging into this (thanks for the trace logs!) and I found the issue (TLDR test case [here](https://github.com/jacksontj/promxy/pull/345)).

The issue here is due to how the anti_affinity config option works. As it stands today the merging will look for a hole of size anti_affinity and then start filling it. This anti_affinity is applied to both sides of the hole -- so we know that it is in fact a hole. So with the config above we'd expect 10s of hole before and after the stitched section. This logic is in place as the general stitching has no guarantee that these are supposed to be the same time range etc. Now in this case the second response had _full_ data -- so no "merging" was needed, we just needed to pick the other series.

The easiest fix for this situation is to start the merge with the larger series (to avoid requiring the merge at all), and I have created  #345 with that change.

Now this does mean if the 2 servers each had a different hold in their data the merged result would still have a hole in it. This is unfortunately very difficult to deal with as we need to leave some space for clock skew etc. I have added a test case to showcase the expected holes in such a case (https://github.com/jacksontj/promxy/pull/345/commits/88c298afcaa27534905b9844bf641b7b9e3124e2) but the fix in the PR should cover this use case :)

**Comment by jacksontj** [comment]
I just merged #345 so this issue should be resolved! Please let me know if you have any other issues or that patch doesn't resolve the issue for you.

**Comment by matejzero** [comment]
Hey there! Great news from your side!

I think your solution will cover my usecases nicely. We mostly have holes during upgrades of software or restarts of the servers. We also have a monitoring node in each datacenter, so in case network connection between DCs is down, at least one server will have full data.

Thanks for the fix!
